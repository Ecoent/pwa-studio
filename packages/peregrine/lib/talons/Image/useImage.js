import { useCallback, useMemo, useState } from 'react';

/**
 * Returns props to render an Image component.
 *
 * @param {function} props.onError callback for error of loading image
 * @param {function} props.onLoad callback for load of image
 * @param {string}   props.unconstrainedSizeKey the key in props.widths for the unconstrained / default width.
 * @param {Map}      props.widths a map of breakpoints to possible widths used to create the img's sizes attribute.
 */
export const useImage = props => {
    const { onError, onLoad, unconstrainedSizeKey, widths } = props;
    const [isLoaded, setIsLoaded] = useState(false);
    const [hasError, setHasError] = useState(false);

    const handleImageLoad = useCallback(() => {
        setIsLoaded(true);

        if (typeof onLoad === 'function') {
            onLoad();
        }
    }, [onLoad]);

    const handleError = useCallback(() => {
        setHasError(true);

        if (typeof onError === 'function') {
            onError();
        }
    }, [onError]);

    // Use the unconstrained / default entry in widths.
    const resourceWidth = useMemo(() => {
        if (!widths) {
            return undefined;
        }

        return widths.get(unconstrainedSizeKey);
    }, [unconstrainedSizeKey, widths]);

    return {
        handleError,
        handleImageLoad,
        hasError,
        isLoaded,
        resourceWidth
    };
};
