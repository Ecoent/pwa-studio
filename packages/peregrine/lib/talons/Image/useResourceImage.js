import { useMemo } from 'react';

/**
 * The talon for working with ResourceImages.
 * Does all the work of generating src, srcSet, and sizes attributes.
 *
 * @param {func}    props.generateSrcset - A function that returns a srcSet.
 * @param {string}  props.resource - The Magento path to the image ex: /v/d/vd12-rn_main_2.jpg
 * @param {number}  props.resourceHeight - The height to request for the fallback image for browsers that don't support srcset / sizes.
 * @param {func}    props.resourceUrl - A function that returns the full URL for the Magento resource.
 * @param {string}  props.type - The Magento image type ("image-category" / "image-product"). Used to build the resource URL.
 * @param {number}  props.width - The width to request for the fallback image for browsers that don't support srcset / sizes.
 * @param {array}   props.widthBreakpoints - The breakpoints related to widths.
 * @param {array}   props.widths - The image widths used by the browser to select the image source.
 */
export const useResourceImage = props => {
    const {
        generateSrcset,
        resource,
        resourceHeight,
        resourceUrl,
        type,
        width,
        widthBreakpoints,
        widths
    } = props;

    const src = useMemo(() => {
        return resourceUrl(resource, {
            type,
            height: resourceHeight,
            width: width
        });
    }, [resource, resourceHeight, resourceUrl, width, type]);

    const srcSet = useMemo(() => {
        return generateSrcset(resource, type);
    }, [generateSrcset, resource, type]);

    // Example: 100px
    // Example: (max-width: 640px) 50px, 100px
    const sizes = useMemo(() => {
        // The values in widths are numbers. Convert to string by adding 'px'.
        const getPixelSize = index => widths[index] + 'px';

        // Helper function for prepending sizes media constraints.
        const constrain = index => {
            const breakpoint = widthBreakpoints[index] + 'px';
            const size = getPixelSize(index);

            return `(max-width: ${breakpoint}) ${size}`;
        };

        // The number of breakpoints must be one less than the number of sizes.
        // The last size entry (the one without a matching breakpoint) is unconstrained.
        const numBreakpoints = widthBreakpoints.length;
        const unconstrainedSizeIndex = numBreakpoints;
        const unconstrainedSize = getPixelSize(unconstrainedSizeIndex);

        // There aren't any breakpoints, this size will always be used.
        if (numBreakpoints === 0) {
            return unconstrainedSize;
        }
        
        // We have some breakpoints. Constrain the sizes with their matching breakpoint.
        // Constrain every size except the last one.
        const widthsToConstrain = widths.slice(0, widths.length);
        const sizesArr = widthsToConstrain.reduce(
            (constrainedSizesArray, _, currentSizeIndex) => {                
                const currentConstraint = constrain(currentSizeIndex);
                constrainedSizesArray.push(currentConstraint);
                return constrainedSizesArray;
            },
            []
        );
        // And add the unconstrained size at the end.
        sizesArr.push(unconstrainedSize);
        
        return sizesArr.join(', ');
    }, [widthBreakpoints, widths]);

    return {
        sizes,
        src,
        srcSet
    };
};
