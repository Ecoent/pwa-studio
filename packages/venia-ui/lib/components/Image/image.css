.root {
    /* For customization, we provide an empty root. */
}

.loaded {
    display: block;
}

.notLoaded {
    display: none;
}

.placeholder_include {
    display: block;
}

.placeholder_exclude {
    display: none;
}
