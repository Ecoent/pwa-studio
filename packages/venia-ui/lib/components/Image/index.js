export { default as Image, UNCONSTRAINED_SIZE_KEY } from './image';
