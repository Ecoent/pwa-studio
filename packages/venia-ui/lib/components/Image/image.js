import React, { Fragment, useMemo } from 'react';
import PropTypes, { func, number, oneOfType, shape, string } from 'prop-types';
import { useImage } from '@magento/peregrine/lib/talons/Image/useImage';
import { transparentPlaceholder } from '@magento/peregrine/lib/util/images';
import { resourceUrl } from '@magento/venia-drivers';

import { generateSrcset } from '../../util/images';
import { mergeClasses } from '../../classify';
import defaultClasses from './image.css';

/**
 * The Image component renders a placeholder until the image is loaded.
 *
 * @param {string}   props.alt the alt text for the image
 * @param {string}   props.classes any classes to apply to this component
 * @param {number}   props.height the desired height of the image
 * @param {function} props.onError callback for error loading image
 * @param {function} props.onLoad callback for when image loads successfully
 * @param {string}   props.placeholder the placeholder source to display while the image loads or if it errors on load
 * @param {string}   props.resource the Magento path to the image ex: /v/d/vd12-rn_main_2.jpg
 * @param {string}   props.sizes the desired sizes attribute of the image
 * @param {string}   props.src the source of the image, ready to use in an img element
 * @param {number}   props.width the desired width of the image
 */
const Image = props => {
    const {
        alt,
        classes: propsClasses,
        onError,
        onLoad,
        placeholder,
        resource,
        resourceHeight,
        resourceWidth,
        sizes,
        src,
        type,
        ...rest
    } = props;

    const talonProps = useImage({
        onError,
        onLoad,
    });

    const {
        handleError,
        handleImageLoad,
        isLoaded,
        shouldRenderPlaceholder
    } = talonProps;

    const classes = mergeClasses(defaultClasses, propsClasses);

    // A placeholder to use until the image is loaded.
    const placeholderClass = shouldRenderPlaceholder ? classes.placeholder_include : classes.placeholder_exclude;
    const placeholderImage = (
        <img
            alt={alt}
            className={placeholderClass}
            loading="eager"
            src={placeholder}
            {...rest}
        />
    );

    /*
     * These don't live in the talon because they depend on @magento/venia-drivers.
     */
    const imageSrcset = useMemo(
        () => generateSrcset(resource, type),
        [resource]
    );
    const source = useMemo(() => {
        // If we have a direct src, use it.
        // Otherwise, get a resourceUrl from the resource and use that.
        return src ? src : resourceUrl(resource, {
            type,
            height: resourceHeight,
            width: resourceWidth
        });
    }, [resourceHeight, type, src, resourceWidth]);

    const imageClass = classes.root + ' ' + (isLoaded ? classes.loaded : classes.notLoaded);
    const actualImage = (
        /*
         * Note: attributes that are allowed to be overridden
         * must appear before the spread of `rest`.
         */
        <img
            loading="lazy"
            {...rest}
            alt={alt}
            className={imageClass}
            onError={handleError}
            onLoad={handleImageLoad}
            src={source}
            srcSet={imageSrcset}
            sizes={sizes}
        />
    );

    return (
        <Fragment>
            {actualImage}
            {placeholderImage}
        </Fragment>
    );
};

const conditionallyRequiredString = (props, propName, componentName) => {
    // This component needs one of src or resource to be provided.
    if (!props.src && !props.resource) {
        return new Error(
            `Missing both 'src' and 'resource' props in ${componentName}. ${componentName} needs at least one of these to be provided.`
        );
    }

    return PropTypes.checkPropTypes({
        resource: string,
        src: string
    }, props, propName, componentName);
};

Image.propTypes = {
    alt: string,
    classes: shape({
        loaded: string,
        notLoaded: string,
        root: string
    }),
    height: oneOfType([number, string]),
    onError: func,
    onLoad: func,
    placeholder: string,
    resource: conditionallyRequiredString,
    resourceHeight: oneOfType([number, string]),
    resourceWidth: oneOfType([number, string]),
    sizes: string,
    src: conditionallyRequiredString,
    type: string,
    width: oneOfType([number, string])
};

Image.defaultProps = {
    placeholder: transparentPlaceholder,
    type: 'image-product'
};

export default Image;
