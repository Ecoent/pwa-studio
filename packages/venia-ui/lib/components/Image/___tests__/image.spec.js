import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useImage } from '@magento/peregrine/lib/talons/Image/useImage';

import Image from '../image';

jest.mock('@magento/peregrine/lib/talons/Image/useImage', () => {
    const useImageTalon = jest.requireActual('@magento/peregrine/lib/talons/Image/useImage');
    const spy = jest.spyOn(useImageTalon, 'useImage');

    return Object.assign(useImageTalon, { useImage: spy });
});

const props = {
    alt: 'Unit Test Image',
    onError: jest.fn(),
    onLoad: jest.fn(),
    placeholder: 'placeholder.jpg',
    src: 'src.jpg',
    fileSrc: 'fileSrc.jpg'
};

const talonProps = {
    handleError: jest.fn(),
    handleImageLoad: jest.fn(),
    hasError: false,
    isLoaded: true,
    shouldRenderPlaceholder: true
};

test('renders a placeholder when appropriate', () => {
    // Arrange.
    useImage.mockReturnValueOnce(talonProps);

    // Act.
    const wrapper = createTestInstance(
        <Image {...props} />
    );

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});

test('renders an image correctly', () => {
    // Arrange.
    const myTalonProps = {
        ...talonProps,
        shouldRenderPlaceholder: false
    };
    useImage.mockReturnValueOnce(myTalonProps);

    // Act.
    const wrapper = createTestInstance(
        <Image {...props} />
    );

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});

test('adds the not loaded class to images that have not been loaded', () => {
    // Arrange.
    const myTalonProps = {
        ...talonProps,
        isLoaded: false
    };
    useImage.mockReturnValueOnce(myTalonProps);

    // Act.
    const wrapper = createTestInstance(
        <Image {...props} />
    );

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});

test('supports overriding the loading attribute', () => {
    // Arrange.
    useImage.mockReturnValueOnce(talonProps);

    // Act.
    const myProps = {
        ...props,
        loading: 'eager'
    };
    const wrapper = createTestInstance(
        <Image {...myProps} />
    );

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});

test('does not generate a srcSet if fileSrc is missing', () => {
    // Arrange.
    useImage.mockReturnValueOnce(talonProps);

    // Act.
    const myProps = {
        ...props,
        fileSrc: undefined
    };
    const wrapper = createTestInstance(
        <Image {...myProps} />
    );

    // Assert.
    expect(wrapper.toJSON()).toMatchSnapshot();
});