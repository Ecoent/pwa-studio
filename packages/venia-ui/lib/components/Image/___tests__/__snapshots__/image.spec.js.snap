// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an image correctly when given src 1`] = `
Array [
  <img
    alt="Unit Test Image"
    className="undefined undefined"
    loading="lazy"
    onError={[MockFunction]}
    onLoad={[MockFunction]}
    src="image_src.jpg"
    srcSet=""
  />,
  <img
    alt="Unit Test Image"
    loading="eager"
    src="placeholder.jpg"
  />,
]
`;
